custom/enikshay/private_sector_datamigration/management/commands/create_dtos_and_agencies.py

```diff
@@ -0,0 +1,102 @@
+
+from corehq.apps.locations.models import SQLLocation, LocationType
+from corehq.apps.locations.tasks import make_location_user
+from custom.enikshay.private_sector_datamigration.models import Agency, UserDetail
+
+
+class Command(BaseCommand):
+
+    def add_arguments(self, parser):
+        parser.add_argument('domain')
+        parser.add_argument('state_code')
+        parser.add_argument('district_code')
+        parser.add_argument('parent_loc_id')
+        parser.add_argument('user_level', choices=['dev', 'real', 'test'])
+        parser.add_argument('org_ids', metavar='org_id', nargs='*', type=int)
+
+    def handle(self, domain, state_code, district_code, parent_loc_id, user_level, org_ids, **options):
+        dto_parent = SQLLocation.active_objects.get(location_id=parent_loc_id)
+        for org_id in org_ids:
+            dto = self.create_dto(domain, state_code, district_code, dto_parent, org_id)
+            for agency in self.get_agencies_by_state_district_org(state_code, district_code, org_id):
+                agency_loc = self.create_agency(domain, agency, dto, org_id)
+                self.create_user(agency_loc, user_level)
+
+    def create_dto(self, domain, state_code, district_code, dto_parent, org_id):
+        return SQLLocation.objects.create(
+            domain=domain,
+            name=self._get_org_name_by_id(org_id),
+            site_code='%s_%s_%d' % (state_code, district_code, org_id),
+            location_type=LocationType.objects.get(
+                domain=domain,
+                code='dto',
+            ),
+            parent=dto_parent,
+            metadata={
+                'enikshay_enabled': 'yes',
+                'is_test': 'no',
+                'private_sector_org_id': org_id,
+            },
+        )
+
+    @staticmethod
+    def get_agencies_by_state_district_org(state_code, district_code, org_id):
+        agency_ids = UserDetail.objects.filter(
+            isPrimary=True,
+        ).filter(
+            districtId=district_code,
+            organisationId=org_id,
+            stateId=state_code,
+        ).values('agencyId').distinct()
+        return Agency.objects.filter(agencyId__in=agency_ids).order_by('agencyId')
+
+    @staticmethod
+    def create_agency(domain, agency, dto, org_id):
+        return SQLLocation.objects.create(
+            domain=domain,
+            name=agency.agencyName,
+            site_code=str(agency.agencyId),
+            location_type=LocationType.objects.get(
+                domain=domain,
+                code=agency.location_type,
+            ),
+            parent=dto,
+            metadata={
+                'enikshay_enabled': 'yes',
+                'is_test': 'no',
+                'nikshay_code': agency.nikshayId,
+                'private_sector_agency_id': agency.agencyId,
+                'private_sector_org_id': org_id,
+
+            },
+        )
+
+    def create_user(self, agency_loc, user_level):
```

Comment:
Yeah, to establish the relationship, you should make sure `location_type.has_user == True`, then 
```python
user.user_location_id = location.location_id
user.set_location(location, commit=False)
user.save()
location.user_id = user._id
location.save()
```
[This](https://github.com/dimagi/commcare-hq/blob/9172c1dcf2c72722047b79a7826496225ab25124/corehq/apps/locations/tasks.py#L101-L112) is the closest existing thing to that.

Probably about time to abstract that in some way too, so feel free if you're so inclined.